"DocFiles.hs: invalid byte sequence" when building hledger with nix

Reported by pop on #hledger. Note stack is using nix in this build because it was enabled in ~/.stack/config.yaml (nix: enabled: true etc.):
```
$ stack --stack-yaml stack8.0.yaml -j 1 build --only-dependencies
hledger-lib-1.4: unregistering (local file changes: Hledger/Data.hs Hledger/Data/Account.hs Hledger/Data/AutoTransaction.hs Hledger/Data/Dates.hs Hle...)
hledger-lib-1.4: configure (lib)
hledger-lib-1.4: build (lib)
hledger-lib-1.4: copy/register
hledger-1.4: configure (lib + exe)
hledger-1.4: build (lib + exe)
Log files have been written to: /home/pop/src/github.com/simonmichael/hledger/.stack-work/logs/
Progress: 1/32
--  While building package hledger-1.4 using:
      /home/pop/.stack/setup-exe-cache/x86_64-linux-nix/Cabal-simple_mPHDZzAJ_1.24.2.0_ghc-8.0.2 --builddir=.stack-work/dist/x86_64-linux-nix/Cabal-1.24.2.0 build lib:hledger exe:hledger --ghc-options " -ddump-hi -ddump-to-file"
    Process exited with code: ExitFailure 1
    Logs have been written to: /home/pop/src/github.com/simonmichael/hledger/.stack-work/logs/hledger-1.4.log

    Configuring hledger-1.4...
    Preprocessing library hledger-1.4...
    [ 1 of 31] Compiling Text.Tabular.AsciiWide ( Text/Tabular/AsciiWide.hs, .stack-work/dist/x86_64-linux-nix/Cabal-1.24.2.0/build/Text/Tabular/AsciiWide.o )
    [ 2 of 31] Compiling Paths_hledger    ( .stack-work/dist/x86_64-linux-nix/Cabal-1.24.2.0/build/autogen/Paths_hledger.hs, .stack-work/dist/x86_64-linux-nix/Cabal-1.24.2.0/build/Paths_hledger.o )
    [ 3 of 31] Compiling Hledger.Cli.Version ( Hledger/Cli/Version.hs, .stack-work/dist/x86_64-linux-nix/Cabal-1.24.2.0/build/Hledger/Cli/Version.o )
    [ 4 of 31] Compiling Hledger.Cli.DocFiles ( Hledger/Cli/DocFiles.hs, .stack-work/dist/x86_64-linux-nix/Cabal-1.24.2.0/build/Hledger/Cli/DocFiles.o )
    
    /home/pop/src/github.com/simonmichael/hledger/hledger/Hledger/Cli/DocFiles.hs:38:6: error:
        * Exception when trying to run compile-time code:
            /home/pop/src/github.com/simonmichael/hledger/hledger/doc/hledger.1: hGetContents: invalid argument (invalid byte sequence)
          Code: (>>=) makeRelativeToProject "doc/hledger.1" embedStringFile
        * In the untyped splice:
            $(makeRelativeToProject "doc/hledger.1" >>= embedStringFile)
```
As far as we know $LANG is set to a valid locale. The locale thing is discussed more at http://hledger.org/manual.html#troubleshooting (may be out of date for GHC 8 ?)

--- a/hledger/Hledger/Cli/DocFiles.hs
+++ b/hledger/Hledger/Cli/DocFiles.hs
@@ -22,6 +22,8 @@ module Hledger.Cli.DocFiles (
 
 import Prelude ()
 import "base-compat-batteries" Prelude.Compat
+import Data.ByteString (ByteString)
+import qualified Data.ByteString.Char8 as BC
 import Data.FileEmbed
 import Data.String
 import System.IO
@@ -32,82 +34,82 @@ import Hledger.Utils (first3, second3, third3)
 
 type Topic = String
 
-docFiles :: IsString a => [(Topic, (a, a, a))]
+docFiles :: [(Topic, (ByteString, ByteString, ByteString))]
 docFiles = [
    ("hledger",
-    ($(makeRelativeToProject "embeddedfiles/hledger.1" >>= embedStringFile)
-    ,$(makeRelativeToProject "embeddedfiles/hledger.txt" >>= embedStringFile)
-    ,$(makeRelativeToProject "embeddedfiles/hledger.info" >>= embedStringFile)
+    ($(makeRelativeToProject "embeddedfiles/hledger.1" >>= embedFile)
+    ,$(makeRelativeToProject "embeddedfiles/hledger.txt" >>= embedFile)
+    ,$(makeRelativeToProject "embeddedfiles/hledger.info" >>= embedFile)
     ))
   ,("hledger-ui",
-    ($(makeRelativeToProject "embeddedfiles/hledger-ui.1" >>= embedStringFile)
-    ,$(makeRelativeToProject "embeddedfiles/hledger-ui.txt" >>= embedStringFile)
-    ,$(makeRelativeToProject "embeddedfiles/hledger-ui.info" >>= embedStringFile)
+    ($(makeRelativeToProject "embeddedfiles/hledger-ui.1" >>= embedFile)
+    ,$(makeRelativeToProject "embeddedfiles/hledger-ui.txt" >>= embedFile)
+    ,$(makeRelativeToProject "embeddedfiles/hledger-ui.info" >>= embedFile)
     ))
   ,("hledger-web",
-    ($(makeRelativeToProject "embeddedfiles/hledger-web.1" >>= embedStringFile)
-    ,$(makeRelativeToProject "embeddedfiles/hledger-web.txt" >>= embedStringFile)
-    ,$(makeRelativeToProject "embeddedfiles/hledger-web.info" >>= embedStringFile)
+    ($(makeRelativeToProject "embeddedfiles/hledger-web.1" >>= embedFile)
+    ,$(makeRelativeToProject "embeddedfiles/hledger-web.txt" >>= embedFile)
+    ,$(makeRelativeToProject "embeddedfiles/hledger-web.info" >>= embedFile)
     ))
   ,("hledger-api",
-    ($(makeRelativeToProject "embeddedfiles/hledger-api.1" >>= embedStringFile)
-    ,$(makeRelativeToProject "embeddedfiles/hledger-api.txt" >>= embedStringFile)
-    ,$(makeRelativeToProject "embeddedfiles/hledger-api.info" >>= embedStringFile)
+    ($(makeRelativeToProject "embeddedfiles/hledger-api.1" >>= embedFile)
+    ,$(makeRelativeToProject "embeddedfiles/hledger-api.txt" >>= embedFile)
+    ,$(makeRelativeToProject "embeddedfiles/hledger-api.info" >>= embedFile)
     ))
   ,("journal",
-    ($(makeRelativeToProject "embeddedfiles/hledger_journal.5" >>= embedStringFile)
-    ,$(makeRelativeToProject "embeddedfiles/hledger_journal.txt" >>= embedStringFile)
-    ,$(makeRelativeToProject "embeddedfiles/hledger_journal.info" >>= embedStringFile)
+    ($(makeRelativeToProject "embeddedfiles/hledger_journal.5" >>= embedFile)
+    ,$(makeRelativeToProject "embeddedfiles/hledger_journal.txt" >>= embedFile)
+    ,$(makeRelativeToProject "embeddedfiles/hledger_journal.info" >>= embedFile)
     ))
   ,("csv",
-    ($(makeRelativeToProject "embeddedfiles/hledger_csv.5" >>= embedStringFile)
-    ,$(makeRelativeToProject "embeddedfiles/hledger_csv.txt" >>= embedStringFile)
-    ,$(makeRelativeToProject "embeddedfiles/hledger_csv.info" >>= embedStringFile)
+    ($(makeRelativeToProject "embeddedfiles/hledger_csv.5" >>= embedFile)
+    ,$(makeRelativeToProject "embeddedfiles/hledger_csv.txt" >>= embedFile)
+    ,$(makeRelativeToProject "embeddedfiles/hledger_csv.info" >>= embedFile)
     ))
   ,("timeclock",
-    ($(makeRelativeToProject "embeddedfiles/hledger_timeclock.5" >>= embedStringFile)
-    ,$(makeRelativeToProject "embeddedfiles/hledger_timeclock.txt" >>= embedStringFile)
-    ,$(makeRelativeToProject "embeddedfiles/hledger_timeclock.info" >>= embedStringFile)
+    ($(makeRelativeToProject "embeddedfiles/hledger_timeclock.5" >>= embedFile)
+    ,$(makeRelativeToProject "embeddedfiles/hledger_timeclock.txt" >>= embedFile)
+    ,$(makeRelativeToProject "embeddedfiles/hledger_timeclock.info" >>= embedFile)
     ))
   ,("timedot",
-    ($(makeRelativeToProject "embeddedfiles/hledger_timedot.5" >>= embedStringFile)
-    ,$(makeRelativeToProject "embeddedfiles/hledger_timedot.txt" >>= embedStringFile)
-    ,$(makeRelativeToProject "embeddedfiles/hledger_timedot.info" >>= embedStringFile)
+    ($(makeRelativeToProject "embeddedfiles/hledger_timedot.5" >>= embedFile)
+    ,$(makeRelativeToProject "embeddedfiles/hledger_timedot.txt" >>= embedFile)
+    ,$(makeRelativeToProject "embeddedfiles/hledger_timedot.info" >>= embedFile)
     ))
   ]
 
 docTopics :: [Topic]
 docTopics = map fst docFiles
 
-lookupDocTxt :: IsString a => Topic -> a
+lookupDocTxt :: Topic -> ByteString
 lookupDocTxt name =
   maybe (fromString $ "No text manual found for topic: "++name) second3 $ lookup name docFiles
 
-lookupDocNroff :: IsString a => Topic -> a
+lookupDocNroff :: Topic -> ByteString
 lookupDocNroff name =
   maybe (fromString $ "No man page found for topic: "++name) first3 $ lookup name docFiles
 
-lookupDocInfo :: IsString a => Topic -> a
+lookupDocInfo :: Topic -> ByteString
 lookupDocInfo name =
   maybe (fromString $ "No info manual found for topic: "++name) third3 $ lookup name docFiles
 
 printHelpForTopic :: Topic -> IO ()
 printHelpForTopic t =
-  putStrLn $ lookupDocTxt t
+  BC.putStr (lookupDocTxt t)
 
 runPagerForTopic :: FilePath -> Topic -> IO ()
 runPagerForTopic exe t = do
   (Just inp, _, _, ph) <- createProcess (proc exe []){
     std_in=CreatePipe
     }
-  hPutStrLn inp (lookupDocTxt t)
+  BC.hPutStrLn inp (lookupDocTxt t)
   _ <- waitForProcess ph
   return ()
 
 runManForTopic :: Topic -> IO ()
 runManForTopic t =
   withSystemTempFile ("hledger-"++t++".nroff") $ \f h -> do
-    hPutStrLn h $ lookupDocNroff t
+    BC.hPutStrLn h $ lookupDocNroff t
     hClose h
      -- the temp file path will presumably have a slash in it, so man should read it
     callCommand $ "man " ++ f
@@ -115,7 +117,7 @@ runManForTopic t =
 runInfoForTopic :: Topic -> IO ()
 runInfoForTopic t =
   withSystemTempFile ("hledger-"++t++".info") $ \f h -> do
-    hPutStrLn h $ lookupDocInfo t
+    BC.hPutStrLn h $ lookupDocInfo t
     hClose h
     callCommand $ "info " ++ f
 

